drivers/flash/flash_ifx_cat1.c

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright 2023 Cypress Semiconductor Corporation (an Infineon company) or
+ * an affiliate of Cypress Semiconductor Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT	  infineon_cat1_flash_controller
+#define SOC_NV_FLASH_NODE DT_PARENT(DT_INST(0, fixed_partitions))
+
+#define PAGE_LEN DT_PROP(SOC_NV_FLASH_NODE, erase_block_size)
+
+#include <zephyr/kernel.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/logging/log.h>
+
+#include "cyhal_flash.h"
+
+LOG_MODULE_REGISTER(flash_infineon_cat1, CONFIG_FLASH_LOG_LEVEL);
+
+/* Device config structure */
+struct ifx_cat1_flash_config {
+	uint32_t base_addr;
+	uint32_t max_addr;
+};
+
+/* Data structure */
+struct ifx_cat1_flash_data {
+	cyhal_flash_t flash_obj;
+	struct k_sem sem;
+};
+
+static struct flash_parameters ifx_cat1_flash_parameters = {
+	.write_block_size = DT_PROP(SOC_NV_FLASH_NODE, write_block_size),
+	.erase_value = 0x00,
+};
+
+static inline void flash_ifx_sem_take(const struct device *dev)
+{
+	struct ifx_cat1_flash_data *data = dev->data;
+
+	k_sem_take(&data->sem, K_FOREVER);
+}
+
+static inline void flash_ifx_sem_give(const struct device *dev)
+{
+	struct ifx_cat1_flash_data *data = dev->data;
+
+	k_sem_give(&data->sem);
+}
+
+static int ifx_cat1_flash_read(const struct device *dev, off_t offset, void *data, size_t data_len)
+{
+	struct ifx_cat1_flash_data *dev_data = dev->data;
+	const struct ifx_cat1_flash_config *dev_config = dev->config;
+	uint32_t read_offset = dev_config->base_addr + offset;
+	cy_rslt_t rslt = CY_RSLT_SUCCESS;
+	int ret = 0;
+
+	flash_ifx_sem_take(dev);
+
+	rslt = cyhal_flash_read(&dev_data->flash_obj, read_offset, (uint8_t *)data, data_len);
+	if (rslt != CY_RSLT_SUCCESS) {
+		LOG_ERR("Error reading @ 0x%x (Err:0x%x)", read_offset, rslt);
+		ret = -EIO;
+		goto out;
+	}
+
+out:
+	flash_ifx_sem_give(dev);
+	return ret;
+}
+
+static int ifx_cat1_flash_write(const struct device *dev, off_t offset, const void *data,
+				size_t data_len)
+{
+	struct ifx_cat1_flash_data *dev_data = dev->data;
+	const struct ifx_cat1_flash_config *dev_config = dev->config;
+	uint32_t write_offset = dev_config->base_addr + (uint32_t)offset;
+	const uint8_t *data_ptr = (const uint8_t *)data;
+	cy_rslt_t rslt = CY_RSLT_SUCCESS;
+	int ret = 0;
+
+	if ((offset < 0) || (data_len % PAGE_LEN != 0) || (offset % PAGE_LEN != 0)) {
+		return -EINVAL;
+	}
+
+	if (data_len == 0) {
+		return 0;
+	}
```

Comment:
Note for future: do short tests first. 
Here if `data_len` is 0, there are 3 other arg tests above it before we even get to this return.
Other thing is that this is 'no-work' test, which means that even if we have bad args to other function parameters, as tested above, there should be no error, because we are not going to do anything anyway, which means that bad args would not harm us anyway.

Reply:
updated to check the data_len first..